File: app/src/main/java/org/mesonet/app/advisories/AdvisoriesFragment.java
package org.mesonet.app.advisories;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import org.mesonet.app.R;
import org.mesonet.app.baseclasses.BaseFragment;
import org.mesonet.app.databinding.AdvisoriesFragmentBinding;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Observable;
import java.util.Observer;

import javax.inject.Inject;



public class AdvisoriesFragment extends BaseFragment implements Observer
{
    private AdvisoriesFragmentBinding mBinding;

    @Inject
    AdvisoryDataProvider mAdvisoryDataProvider;

    @Inject
    AdvisorySorter mAdvisorySorter;



    @Override
    public View onCreateView(LayoutInflater inInflater, ViewGroup inParent, final Bundle inSavedInstanceState)
    {
        mBinding = DataBindingUtil.inflate(inInflater, R.layout.advisories_fragment, inParent, false);

        mAdvisoryDataProvider.addObserver(this);
        mBinding.advisoriesRecyclerView.setAdapter(new AdvisoriesRecyclerViewAdapter());

        return mBinding.getRoot();
    }



    @Override
    public void update(Observable o, Object arg)
    {
        List<AdvisoryModel> advisories = mAdvisoryDataProvider.GetAdvisories();

        if(advisories != null)
        {
            List<Object> endResult = new ArrayList<>();

            AdvisoryModel.AdvisoryLevel currentLevel = null;
            AdvisoryModel.AdvisoryType currentType = null;

            advisories = mAdvisorySorter.Sort(advisories);

            for(int i = 0; i < advisories.size(); i++)
            {
                if(advisories.get(i).mAdvisoryType.mAdvisoryLevel != currentLevel ||
                        advisories.get(i).mAdvisoryType.mAdvisoryType != currentType)
                {
                    endResult.add(advisories.get(i).mAdvisoryType);
                    currentLevel = advisories.get(i).mAdvisoryType.mAdvisoryLevel;
                    currentType = advisories.get(i).mAdvisoryType.mAdvisoryType;
                }

                endResult.add(advisories.get(i));
            }

            mBinding.advisoriesRecyclerView.SetItems(endResult);
        }
    }
}

File: app/src/main/java/org/mesonet/app/advisories/AdvisoriesFragment.kt
package org.mesonet.app.advisories

import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import org.mesonet.app.R

import org.mesonet.app.baseclasses.BaseFragment
import org.mesonet.app.databinding.AdvisoriesFragmentBinding
import org.mesonet.dataprocessing.PageStateInfo
import org.mesonet.dataprocessing.advisories.AdvisoryDataProvider
import org.mesonet.dataprocessing.advisories.AdvisoryDisplayListBuilder


import javax.inject.Inject


class AdvisoriesFragment : BaseFragment()
{
    private var mBinding: AdvisoriesFragmentBinding? = null

    @Inject
    internal lateinit var mAdvisoryDataProvider: AdvisoryDataProvider

    @Inject
    internal lateinit var mAdvisoryListBuilder: AdvisoryDisplayListBuilder

    var mAdvisoryDisposable: Disposable? = null
    var mPageStateDisposable: Disposable? = null



    override fun onCreateView(inInflater: LayoutInflater, inParent: ViewGroup?, inSavedInstanceState: Bundle?): View? {
        mBinding = DataBindingUtil.inflate(inInflater, R.layout.advisories_fragment, inParent, false)

        mBinding?.advisoriesRecyclerView?.setAdapter(AdvisoriesRecyclerViewAdapter())

        return mBinding?.root
    }


    override fun onResume() {
        super.onResume()

        if(mPageStateDisposable?.isDisposed != false) {
            mAdvisoryDataProvider.GetPageStateObservable().observeOn(AndroidSchedulers.mainThread()).subscribe(object : Observer<PageStateInfo> {
                override fun onComplete() {}

                override fun onSubscribe(d: Disposable) {
                    mPageStateDisposable = d
                }

                override fun onNext(t: PageStateInfo) {
                    when(t.GetPageState())
                    {
                        PageStateInfo.PageState.kData ->
                        {
                            (mBinding?.advisoriesRecyclerView as RecyclerView?)?.visibility = View.VISIBLE
                            mBinding?.errorText?.visibility = View.GONE
                            mBinding?.progressBar?.visibility = View.GONE
                        }
                        PageStateInfo.PageState.kLoading ->
                        {
                            (mBinding?.advisoriesRecyclerView as RecyclerView?)?.visibility = View.GONE
                            mBinding?.errorText?.visibility = View.GONE
                            mBinding?.progressBar?.visibility = View.VISIBLE
                        }
                        PageStateInfo.PageState.kError ->
                        {
                            (mBinding?.advisoriesRecyclerView as RecyclerView?)?.visibility = View.VISIBLE
                            mBinding?.errorText?.text = t.GetErrorMessage()
                            mBinding?.errorText?.visibility = View.VISIBLE
                            mBinding?.progressBar?.visibility = View.GONE
                        }
                    }
                }

                override fun onError(e: Throwable) {
                    e.printStackTrace()
                }

            })
        }

        if(mAdvisoryDisposable?.isDisposed != false) {
            mAdvisoryDataProvider.GetDataObservable().observeOn(AndroidSchedulers.mainThread()).flatMap { list ->
                mAdvisoryListBuilder.BuildList(list)
            }.observeOn(AndroidSchedulers.mainThread()).subscribe(object : Observer<ArrayList<Pair<AdvisoryDisplayListBuilder.AdvisoryDataType, AdvisoryDisplayListBuilder.AdvisoryData>>> {
                override fun onComplete() {}
                override fun onSubscribe(d: Disposable) {
                    mAdvisoryDisposable = d
                }

                override fun onError(e: Throwable) {
                    e.printStackTrace()
                    if (mBinding != null) {
                        mBinding?.progressBar?.visibility = View.GONE
                        mBinding?.errorText?.visibility = View.VISIBLE
                    }
                }

                override fun onNext(t: ArrayList<Pair<AdvisoryDisplayListBuilder.AdvisoryDataType, AdvisoryDisplayListBuilder.AdvisoryData>>) {
                    if (mBinding != null) {
                        mBinding?.progressBar?.visibility = View.GONE
                        mBinding?.advisoriesRecyclerView?.SetItems(t)

                        if (t.size == 0)
                            mBinding?.errorText?.visibility = View.VISIBLE
                        else
                            mBinding?.errorText?.visibility = View.GONE
                    }
                }
            })
        }
    }


    override fun onPause() {
        mAdvisoryDisposable?.dispose()
        mAdvisoryDisposable = null
        mPageStateDisposable?.dispose()
        mPageStateDisposable = null
        super.onPause()
    }
}
